File: construct-binary-tree-from-preorder-and-inorder-traversal.py
from typing import Optional, Self

class TreeNode:
    def __init__(self, val: int = 0, left: Optional[Self] = None, right: Optional[Self] = None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if len(preorder) == 0 or len(inorder) == 0: return None

        imap = {val: i for i, val in enumerate(inorder)}

        def build(pstart: int, pend: int, istart: int, iend: int) -> TreeNode:
            if pend < pstart or iend < istart: return None

            root_val = preorder[pstart]
            root = TreeNode(root_val)
            
            imid = imap[root_val]
            left_len = imid - istart

            root.left = build(pstart+1, pstart+left_len, istart, istart + left_len)
            root.right = build(pstart+left_len+1, pend, imid+1, iend)

            return root

        return build(0, len(preorder)-1, 0, len(inorder)-1)

File: construct-binary-tree-from-preorder-and-inorder-traversal.cpp
#include <vector>
#include <unordered_map>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int, int> imap;
		for (int i=0; i<inorder.size(); ++i) imap[inorder[i]] = i;

		return build(0, preorder.size()-1, 0, inorder.size()-1, preorder, imap);
    }

private:
	TreeNode* build(int pstart, int pend, int istart, int iend, vector<int>& preorder, unordered_map<int, int>& imap) {
		if ((pend < pstart) || (iend < istart)) return nullptr;

		int val = preorder[pstart];
		TreeNode *root = new TreeNode(val);
		
		int imid = imap[val];
		int left_len = imid - istart;

		root->left = build(pstart+1, pstart+left_len, istart, istart + left_len, preorder, imap);
		root->right = build(pstart+left_len+1, pend, imid+1, iend, preorder, imap);

		return root;
	}	
};
